src/custom/state/appData/atoms.ts

```diff
@@ -0,0 +1,96 @@
+import { atom } from 'jotai'
+import { atomWithStorage } from 'jotai/utils'
+
+import {
+  AddAppDataToUploadQueueParams,
+  AppDataPendingToUpload,
+  AppDataInfo,
+  FlattenedAppDataFromUploadQueue,
+  RemoveAppDataFromUploadQueueParams,
+  UpdateAppDataOnUploadQueueParams,
+} from 'state/appData/types'
+import { buildAppDataRecordKey, parseAppDataRecordKey } from 'state/appData/utils'
+
+const INITIAL_APP_DATA_UPLOAD_QUEUE: AppDataPendingToUpload = {}
+
+/**
+ * Base atom that store the current appDataInfo
+ */
+export const appDataInfoAtom = atom<AppDataInfo | null>(null)
+
+/**
+ * Base atom that stores all appData pending to be uploaded
+ */
+export const appDataUploadQueueAtom = atomWithStorage(
+  'appDataUploadQueue', // local storage key
+  INITIAL_APP_DATA_UPLOAD_QUEUE
+)
+
+/**
+ * Write only atom to add an appData to upload queue
+ */
+export const addAppDataToUploadQueueAtom = atom(
+  null,
+  (get, set, { chainId, orderId, appData }: AddAppDataToUploadQueueParams) => {
+    set(appDataUploadQueueAtom, () => {
+      const docs = get(appDataUploadQueueAtom)
+      const key = buildAppDataRecordKey({ chainId, orderId })
+
+      return {
+        ...docs,
+        [key]: { ...appData, uploading: false },
+      }
+    })
+  }
+)
+
+/**
+ * Write only atom to update upload status of an appData on upload queue
+ */
+export const updateAppDataOnUploadQueueAtom = atom(
+  null,
+  (get, set, { chainId, orderId, uploading, tryAfter }: UpdateAppDataOnUploadQueueParams) => {
+    set(appDataUploadQueueAtom, () => {
+      const docs = get(appDataUploadQueueAtom)
+      const key = buildAppDataRecordKey({ chainId, orderId })
+
+      if (!docs[key]) {
+        return docs
+      }
+
+      return {
+        ...docs,
+        [key]: { ...docs[key], uploading, tryAfter },
+      }
```

Comment:
isn't this update function just a more general version of what u created in `addAppDataToUploadQueueAtom`. If this doesn't add any complication, I  would refactor and then make the other call this one with `tryAfter = false`

Reply:
as i said, only if u think is easy, its not that bad in this case, and its quite readable

Reply:
In my opinion having two separate functions for `add` and `update` logic makes sense even though there is not too much difference in this case, but I would keep them separated. Also as you said its quite readable.

Reply:
The interface is actually different (appData on add, uploading and tryAfter on udpate), and I rather keep them separated like Nenad says.

To have the same atom handling all cases means we are back at a sort of reducer which is what I want to avoid in the first place :)